src/Rector/Deprecation/ActionAnnotationToAttributeRector.php

```diff
@@ -0,0 +1,171 @@
+
+/**
+ * Publishes an entity.
+ *
+ * @Action(
+ *   id = "entity:publish_action",
+ *   action_label = @Translation("Publish"),
+ *   deriver = "Drupal\Core\Action\Plugin\Action\Derivative\EntityPublishedActionDeriver",
+ * )
+ */
+class PublishAction extends EntityActionBase {
+CODE_SAMPLE
+, <<<'CODE_SAMPLE'
+
+namespace Drupal\Core\Action\Plugin\Action;
+
+use Drupal\Core\Action\Plugin\Action\Derivative\EntityPublishedActionDeriver;
+use Drupal\Core\Action\Attribute\Action;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\Core\StringTranslation\TranslatableMarkup;
+
+/**
+ * Publishes an entity.
+ */
+#[Action(
+  id: 'entity:publish_action',
+  action_label: new TranslatableMarkup('Publish'),
+  deriver: EntityPublishedActionDeriver::class
+)]
+class PublishAction extends EntityActionBase {
+CODE_SAMPLE
+)]);
+    }
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes() : array
+    {
+        return [Class_::class];
+    }
+    public function provideMinPhpVersion() : int
+    {
+        return PhpVersion::PHP_81;
+    }
+    /**
+     * @param Class_|ClassMethod $node
+     */
+    public function refactor(Node $node) : ?Node
+    {
+        $phpDocInfo = $this->phpDocInfoFactory->createFromNode($node);
+        if (!$phpDocInfo instanceof PhpDocInfo) {
+            return null;
+        }
+        /** @var PhpDocTagNode[] $tagsByName */
+        $tagsByName = $phpDocInfo->getTagsByName('Action');
+        if ($tagsByName === []) {
+            return null;
+        }
+        $hasChanged = \false;
+        foreach ($tagsByName as $valueNode) {
+            if (!$valueNode->value instanceof GenericTagValueNode) {
+                continue;
+            }
+            $stringValue = $valueNode->value->value;
+            $stringValue = '{' . trim($stringValue, '()') . '}';
+            $tokenIterator = $this->tokenIteratorFactory->create($stringValue);
+            $data = $this->arrayParser->parseCurlyArray($tokenIterator, $node);
+            $attribute = $this->createAttribute($data);
+            $node->attrGroups[] = new AttributeGroup([$attribute]);
+            // cleanup
+            $this->phpDocTagRemover->removeTagValueFromNode($phpDocInfo, $valueNode);
+            $hasChanged = \true;
+        }
+        if ($hasChanged) {
+            $this->docBlockUpdater->updateRefactoredNodeWithPhpDocInfo($node);
+            return $node;
+        }
+        return null;
+    }
+    private function createAttribute(array $parsedArgs) : Attribute
+    {
+        $fullyQualified = new FullyQualified(Action::class);
+        $args = [];
+        foreach ($parsedArgs as $value) {
+            if ($value->key === 'label') {
+                $arg = new Node\Expr\New_(new Node\Name(TranslatableMarkup::class), [new Arg(new String_($value->value->values[0]->value->value))]);
+            }
+            else {
+                $arg = new String_($value->value->value);
```

Comment:
In our internal projects we tend to use a `const PLUGIN_ID = 'foo'` pattern and reference these in our annotations. I've been playing around with a custom annotation to attribute rector rule based on this one and noticed it fails on these constants. The following seems to work, although I'm wondering if there is a better way to detect the use of a constant in the annotation.

```
$arg = new String_($value->value instanceof StringNode ? new String_($value->value->value) : new ConstFetch(new Node\Name($value->value)));
```

Reply:
See root comment.